src/mpid/ch4/shm/posix/posix_am.h

```diff
@@ -3,114 +3,401 @@
-                                     MPI_Count count,
-                                     MPI_Datatype datatype, MPIR_Request * sreq, void *shm_context)
+#include <posix_eager_impl.h>
+
+#undef FUNCNAME
+#define FUNCNAME MPIDI_POSIX_am_isend
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int
+MPIDI_POSIX_am_isend(int rank,
+                     MPIR_Comm * comm,
+                     int handler_id,
+                     const void *am_hdr,
+                     size_t am_hdr_sz,
+                     const void *data,
+                     MPI_Count count, MPI_Datatype datatype, MPIR_Request * sreq, void *shm_context)
 {
-    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_POSIX_AM_ISEND);
-    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_POSIX_AM_ISEND);
+    int mpi_errno = MPI_SUCCESS;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_POSIX_SEND_AM);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_POSIX_SEND_AM);
+
+    int result = MPIDI_POSIX_OK;
+
+    int dt_contig;
+    size_t data_sz;
+    MPIR_Datatype *dt_ptr;
+    MPI_Aint dt_true_lb;
+
+    uint8_t *send_buf = NULL;
+
+    MPIDI_Datatype_get_info(count, datatype, dt_contig, data_sz, dt_ptr, dt_true_lb);
+
+    send_buf = (uint8_t *) data + dt_true_lb;
+
+    MPIDI_POSIX_am_header_t msg_hdr = {.handler_id = handler_id,
+        .am_hdr_sz = am_hdr_sz,
+        .data_sz = data_sz
+    };
+#ifdef POSIX_AM_DEBUG
+    static int seq_num = 0;
+    msg_hdr.seq_num = seq_num++;
+#endif /* POSIX_AM_DEBUG */
+
+    MPIDI_POSIX_am_header_t *msg_hdr_p = &msg_hdr;
+    MPIDI_POSIX_am_request_header_t *curr_sreq_hdr = NULL;
+
+    if (unlikely(!dt_contig)) {
+        size_t segment_first;
+        MPI_Aint last;
+        struct MPIDU_Segment *segment_ptr = NULL;
+
+        segment_ptr = MPIDU_Segment_alloc();
+
+        MPIR_ERR_CHKANDJUMP1(segment_ptr == NULL, mpi_errno,
+                             MPI_ERR_OTHER, "**nomem", "**nomem %s", "Send MPIDU_Segment_alloc");
+
+        MPIDU_Segment_init(data, count, datatype, segment_ptr, 0);
+
+        segment_first = 0;
+        last = data_sz;
+
+        MPIDI_POSIX_AMREQUEST(sreq, req_hdr) = NULL;
+
+        /* Prepare private storage */
+
+        mpi_errno = MPIDI_POSIX_am_init_request(am_hdr, am_hdr_sz, sreq);
+        if (mpi_errno)
+            MPIR_ERR_POP(mpi_errno);
+
+        curr_sreq_hdr = MPIDI_POSIX_AMREQUEST(sreq, req_hdr);
+
+        curr_sreq_hdr->pack_buffer = (char *) MPL_malloc(data_sz);
+
+        MPIR_ERR_CHKANDJUMP1(MPIDI_POSIX_AMREQUEST_HDR(sreq, pack_buffer) == NULL, mpi_errno,
+                             MPI_ERR_OTHER, "**nomem", "**nomem %s", "Send Pack buffer alloc");
+
+        MPIDU_Segment_pack(segment_ptr, segment_first, &last,
+                           MPIDI_POSIX_AMREQUEST_HDR(sreq, pack_buffer));
+        MPIDU_Segment_free(segment_ptr);
+
+        send_buf = (uint8_t *) curr_sreq_hdr->pack_buffer;
+    }
+
+    const int grank = MPIDI_CH4U_rank_to_lpid(rank, comm);
```

Comment:
@ddurnov I think this is another problem. We are using essentially the MPI_COMM_WORLD rank where we should be using node local rank. As it is, it works fine for single node jobs, but multinode (or debug mode) will segfault if the rank index overruns the allocated segment. @yfguo we need an efficient mechanism to translate MPI_COMM_WORLD rank to node local rank in order for this work.